### PR TITLE
[1561] Incorrect feed sorting

### LINKED ISSUE #1561: -asc sort orders reverse in-group sorting
**Newsboat version (copy the output of `newsboat -v` or the first line of `git show`)**: 2.23-15-g3cd0

Config file (copy from ~/.newsboat/config or ~/.config/newsboat/config):

```
feed-sort-order unreadarticlecount-asc
```

Steps to reproduce the issue:

1. Have a dozen feeds in the urls file.

2. Press `R` to reload them.

3. Note down the order, than press `R` to reload again.

**Other info you think is relevant**:

This is quite hard to explain, but I'll try.

Suppose we have five feeds, where first column is the number of unread articles, and the second column in the title of the feed:

```
3 a
2 b
1 c
1 d
1 e
```

These are in "unreadarticlecount-asc" order.

Now let's mentally execute `FeedContainer::sort_feeds` on this list. First, [we run `stable_sort` that puts feeds with fewer articles at the beginning](https://github.com/newsboat/newsboat/blob/3cd079beac3cb312820d26d618e2e759bc727637/src/feedcontainer.cpp#L59-L66). The result is:

```
1 c
1 d
1 e
2 b
3 a
```

Note that the relative order of "c", "d", and "e" is preserved by the stable sort. Now we run `reverse` to put this into "asc" sort:

```
3 a
2 b
1 e
1 d
1 c
```

Note how the relative order of "c", "d", and "e" is now reversed. If we execute `sort_feeds` again, these three feeds will get back into their correct order.

Reported by Xianx on IRC.

### PR DESCRIPTION
**Pull request by tau3**
Hello!

This is my attempt to fix #1561 

Making a corner-case for "unread article count" method in a separate class may be not the best ever approach, but this is why we have code reviews :)

### CODE REVIEW

**Comment by Minoru** [comment]
Oh, thanks for taking a stab at that issue! It's also nice you wrote the tests right away; it usually takes a reminder before they're added :)

> Making a corner-case for "unread article count" method in a separate class may be not the best ever approach

Yeah :) I see two problems:

1. the "unread article count" is not actually special. #1561 talks about it because it's how the issue was reported, but any sort where ties can occur is affected — and that's all of them;
2. when sorting by "unread article count", this PR resolves ties by "title" field. This might not be what users want, and they have no way to override it. Previously, they could just sort by the secondary field first, and then by the primary; this worked as expected due to the magic of stable sorting. I'd prefer to keep this behaviour. Leave the ties be. (There's an old Newsbeuter issue about sorting by multiple fields at once, https://github.com/akrennmair/newsbeuter/issues/467. You can address it separately if that's something you'd like to work on.)

With all of that said, I still think that the core idea of this PR is sound: the comparator function should look at the sorting direction, and flip between `<` and `>`. Now that I look at it, `RssFeed::sort_unlocked` does that too. I think that's the correct approach.

Finally, it appears that the PR is based on some older history, and contains merge commits. That makes it harder to review: instead of looking at just the two new commits, I'm seeing some stuff that is already merged into `master`. If I were to merge this PR, the history will look even worse >_< When in the future someone stumbles upon this while bisecting, they'll be thoroughly confused. Since it looks like you'll start afresh with this PR, just `git reset --hard master`, make new commits and then `git push --force`, please.

**Comment by tau3** [comment]
Thanks for review, here are the changes. Unfurtunately, I can't run unit tests on MacOS after recent merge from master (some rust linking issue), but everything should work.

**Comment by dennisschagt** [comment]
> I can't run unit tests on MacOS after recent merge from master (some rust linking issue)

Could you try removing `src/logger.o` and then rerunning the tests?

**Comment by tau3** [comment]
@dennisschagt, thank you for advice!
`All tests passed (23068 assertions in 547 test cases)`

File: src/feedcontainer.cpp [hunk 1]
```diff
@@ -9,95 +9,120 @@
...
+	{
+		int diff = a->unread_item_count() - b->unread_item_count();
+		bool result = false;
+		if (diff != 0) {
+			if (sort_strategy.sd == SortDirection::DESC) {
+				diff *= -1;
+			}
+			result = diff > 0;
+		}
+		return result;
```
Comment (Minoru) [inline src/feedcontainer.cpp:30]: Can this code be moved into `sort_feeds`? It's a bit weird to have a separate class for this when we can have it inline (like we do for all other sorts).

Also, I think `diff` is superfluous. Why not initialize `result` directly, like in other sorts?

File: test/feedcontainer.cpp [hunk 2]
```diff
@@ -382,6 +382,60 @@ TEST_CASE("sort_feeds() sorts by number of articles in a feed "
 	}
 }
 
+TEST_CASE("sort_feeds() and keep in-group order", "[FeedContainer]")
```
Comment (Minoru) [inline test/feedcontainer.cpp:385]: Can this test please be extended to other sorting fields?

**Review by Minoru — CHANGES_REQUESTED** [review]
Looks pretty good! I think this needs one last push, and then it'll be ready for merging.

Thank you for taking the time to do this, keep up the good work :+1:

File: src/feedcontainer.cpp (hunk 1 above)
Comment (tau3) [reply src/feedcontainer.cpp]: because result depends on three possible states of diff, not two (like boolean). maybe it cant be done better :thinking:

File: test/feedcontainer.cpp (hunk 2 above)
Comment (tau3) [reply test/feedcontainer.cpp:385]: do you mean a test for each feed sort method?

File: src/feedcontainer.cpp (hunk 1 above)
Comment (Minoru) [reply src/feedcontainer.cpp]: I don't see a reason to special-case `diff == 0`. The line `result = diff > 0` already handles this case correctly, returning `false`; the change of sort strategy doesn't change the result, because `0 * -1 == 0`. So I'm pretty sure you can drop the `if` and initialize the `result` right away.

File: test/feedcontainer.cpp (hunk 2 above)
Comment (Minoru) [reply test/feedcontainer.cpp:385]: Yes. Since they're implemented separately, we got to have a test for each. If we don't do this now, then sometime in the future someone might change something in one of the methods, see that tests pass, and conclude that everything's fine, even though they might've broken the results of this PR.

**Review by Minoru — APPROVED** [review]
Looks good to me. Thank you for all the work, @tau3! Merging.
